### ISSUE TITLE
How to improve accuracy?

### ISSUE THREAD

**Issue by tunglambk** [YangLin-George/PLELog#13]
Hi Dr @YangLin-George ,

PLELog is applied to reduce the cost of labeling data. But if I use PLELog and after a long time, I have more data that I know if they are normal or abnormal, and there are logs that PLELog detects wrongly. How can I improve it?

Thank you so much.

**Comment by YangLin-George** [comment]
Hi @tunglambk , Thanks for the question!

If you have known normal and anomalous log data, you can directly **RETRAIN Attention-based GRU** part to increase its accuracy. 

You can refer to the `Instance` class in `entities/instances.py` as a reference to construct new training instances. You may create a new list of training instances as shown in [line#159](https://github.com/YangLin-George/PLELog/blob/fd35d79c9bd38b2d087e2241d5a942113013640a/approaches/PLELog.py#L159) in `approaches/PLELog.py` and follow the training process afterward. 

Please note if training PLELog under the supervision mode, the `confidence` in instances should be set to 0 as default, and the `label`should be the correct label (Normal or Anomalous, in a string).

That should answer your question and I hope it is useful to you. If anything, please let me know.

Lin.

**Comment by tunglambk** [comment]
Hi Dr @YangLin-George ,
Thank you for the answer. I also have other question.
I can evaluate the result when I train with HDFS dataset/BGL dataset because both of them are labelled. But if I use an unlabeled dataset, how I can know when to stop training the model or choose which model at which epoch. I notice that, in many cases, the f1 score fluctuates significantly. This is an example of f1 score in 5 epochs.

`[98.29426653460987,
 92.18523878437048,
 98.79384751576852,
 94.01493636268012,
 93.87583008327184]`

Thank you so much.

**Comment by YangLin-George** [comment]
If the log data are completely unlabeled, I think the loss values could be a possible sign of when to stop. However, the results of the 5 epoches show that the it does not converge yet, maybe you can raise the number of training epoches to see if it will reach a rather stable result.

**Comment by tunglambk** [comment]
Hi Dr @YangLin-George ,
I think loss value is not a good sign to decide. This is the loss of the above example.
`
[0.2339958442194808,
 0.19907579948544823,
 0.18674850066302603,
 0.17882776495710176,
 0.17186708890542549]`

The loss decrease after each epoch but f1 score does not.

Besides that, I trained the model with more epochs (sorry but I lost the training history in that case), and the f1 score when the model converges is not the best.

Thank you so much.

**Comment by YangLin-George** [comment]
Hi @tunglambk ,
As far as I know, the validation set is designed to make sure the model does not overfit while the loss curve is an evidence that  the model fits the training data. These are contradictive purposes, so loss curve sometimes does not work. I think that somhow explains the results you have. 

But again, if validation set is un-available, it is a siginificant challenge to decide when to stop training. I am not an expert in deep learning, perhaps there is a solution to that in other research fields. I'm sorry that I can't help you on this problem. 

Have a nice day!

**Comment by tunglambk** [comment]
Thank you so much Dr @YangLin-George.
